extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/controller/ControllerUtils.java

```diff
@@ -0,0 +1,334 @@
+    final Set<StageId> stagesRun = new HashSet<>();
+
+    while (stagesRun.size() < queryDef.getStageDefinitions().size()) {
+      // 1) Run all stages that cannot stream their output, as solo groups.
+      boolean didRun;
+      do {
+        didRun = false;
+
+        for (final StageId stageId : ImmutableList.copyOf(inflow.keySet())) {
+          if (!stagesRun.contains(stageId)
+              && inflow.get(stageId).isEmpty()
+              && !canStreamOutput(queryDef, stageId.getStageNumber(), config, outflow)) {
+            stagesRun.add(stageId);
+            stageGroups.add(
+                new StageGroup(
+                    Collections.singletonList(stageId),
+                    getOutputChannelMode(
+                        queryDef,
+                        stageId.getStageNumber(),
+                        destination.toSelectDestination(),
+                        useDurableStorage,
+                        false
+                    )
+                )
+            );
+
+            removeStageFlow(stageId, inflow, outflow);
+            didRun = true;
+          }
+        }
+      } while (didRun);
+
+      // 2) Pick some stage that can stream its output, and run that as well as all ready-to-run dependents.
+      StageId currentStageId = null;
+      for (final StageId stageId : ImmutableList.copyOf(inflow.keySet())) {
+        if (!stagesRun.contains(stageId)
+            && inflow.get(stageId).isEmpty()
+            && canStreamOutput(queryDef, stageId.getStageNumber(), config, outflow)) {
+          currentStageId = stageId;
+          break;
+        }
+      }
+
+      if (currentStageId != null) {
+        final List<StageId> currentStageGroup = new ArrayList<>();
+        final int maxStageGroupSize;
+        if (stageGroups.isEmpty()) {
+          maxStageGroupSize = config.getMaxConcurrentStages();
+        } else {
+          final StageGroup priorGroup = stageGroups.get(stageGroups.size() - 1);
+          if (priorGroup.lastStageOutputChannelMode() == OutputChannelMode.MEMORY) {
+            // Prior group must run concurrently with this group.
+            maxStageGroupSize = config.getMaxConcurrentStages() - priorGroup.size();
+          } else {
+            // Prior group can exit before this group starts.
+            maxStageGroupSize = config.getMaxConcurrentStages();
+          }
+        }
+
+        OutputChannelMode currentOutputChannelMode = null;
+        while (currentStageId != null) {
+          final boolean canStream = canStreamOutput(queryDef, currentStageId.getStageNumber(), config, outflow);
+          final Set<StageId> currentOutflow = outflow.get(currentStageId);
+
+          final int maxStageGroupSizeAllowingForDownstreamConsumer;
+          if (queryDef.getStageDefinition(currentStageId).doesSortDuringShuffle()) {
+            // When the current group sorts, there's a pipeline break, so we can "leapfrog": close the prior group
```

Comment:
There's this notion of a group sorting, that is tied to the stage's sorting. 

When is a group considered to be sorting, when all, first, or the last of the stages sort? Logically, it should be first, but there's no indication in the code that implies that.

Reply:
The final stage is the only one that can possibly sort, because of the check below that closes off the stage group if currentStageId "doesSortDuringShuffle()". I think the logic here would apply if _any_ stage in the group sorted, though. (If any stage sorts, that causes a pipeline break that means the prior stage group would no longer be needed.)